File: lab/python/arrays_e_listas/produto_escalar.py
def produto_escalar(v1, v2):
    if len(v1) != len(v2):
        raise ValueError("Vetores precisam ter o mesmo tamanho")
    soma = 0
    for i in range(len(v1)):
        soma += v1[i] * v2[i]
    return soma


def produto_escalar2(v1, v2):
    return sum(a*b for a, b in zip(v1, v2))


print(produto_escalar([1, 2, 3], [4, 5, 6]))  # 32
print(produto_escalar2([1, 2, 3], [4, 5, 6]))  # 32

File: lab/c/arrays_e_listas/produto_escalar.c
#include <stdio.h>
#include <stdlib.h>

double produto_escalar(double *v1, double *v2, int n1, int n2) {
    if (n1 != n2) {
        fprintf(stderr, "Erro: Vetores precisam ter o mesmo tamanho\n");
        exit(1);
    }

    double soma = 0;
    for (int i = 0; i < n1; i++) {
        soma += v1[i] * v2[i];
    }
    return soma;
}

double produto_escalar2(double *v1, double *v2, int n1, int n2) {
    if (n1 != n2) {
        fprintf(stderr, "Erro: Vetores precisam ter o mesmo tamanho\n");
        exit(1);
    }

    double soma = 0;
    for (int i = 0; i < n1; i++) {
        soma += v1[i] * v2[i];
    }
    return soma;
}

int main(void) {
    double v1[] = {1, 2, 3};
    double v2[] = {4, 5, 6};
    int n = sizeof(v1) / sizeof(v1[0]);

    printf("%.0f\n", produto_escalar(v1, v2, n, n));
    printf("%.0f\n", produto_escalar2(v1, v2, n, n));

    return 0;
}
